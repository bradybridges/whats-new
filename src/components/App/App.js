import React, { Component } from 'react';
import './App.scss';
import NewsContainer from '../NewsContainer/NewsContainer';
import Menu from '../Menu/Menu';
import SearchForm from '../SearchForm/SearchForm';
import local from '../../data/local';

class App extends Component {
  constructor() {
    super();
    this.state = {
      news: null,
      currentCategory: 'local',
      currentNews: local
    }
    this.fetchData();
  }

  fetchData = () => {
    fetch('https://whats-new-api.herokuapp.com/api/v1/news')
    .then(data => data.json())
    .then(json => this.setState({news: json}))
    .catch(error => console.log(error));
  }

  updateCurrentCategory = (category) => {
    this.setState({currentCategory: category});
    this.updateCurrentNews(category);
  }

  updateCurrentNews = (category) => {
    const updatedNews = this.state.news[category];
    this.setState({currentNews: updatedNews});
  }

  updateNews = (event) => {
    const searchInput = event.target.value.toLowerCase();
    const currentCategory = this.state.currentCategory;
    const updatedNews = this.state.news[currentCategory].filter(news => {
      const headline = news.headline.toLowerCase();
      const description = news.description.toLowerCase();
      if(headline.includes(searchInput) || description.includes(searchInput)) {
        return true;
      } else {
        return false;
      }
    });
    this.setState({currentNews: updatedNews});
  }

  render () {
    return (
      <main>
        <header>
          <h1>Denver News</h1>
          <SearchForm updateNews={this.updateNews} />
        </header>
        <Menu updateCurrentCategory={this.updateCurrentCategory}/>
        <NewsContainer 
          news={this.state.currentNews} 
          currentCategory={this.state.currentCategory}
        />
      </main>
    );
  }
}

export default App;
