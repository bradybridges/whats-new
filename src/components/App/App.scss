$primary-color: #171618fa;
$secondary-color: #e3eef0;
$tertiary-color: #779c43;

* {
  box-sizing: border-box;
  margin: 0;
  padding: 0;
}

body {
  font-family: 'Roboto Slab', serif;
}

header {
  align-items: center;
  background-color: $primary-color;
  color: $tertiary-color;
  display: flex;
  height: 25vh;
  justify-content: space-between;
}

h1 {
  font-size: 5em;
  margin-left: 64px;
}

.app {
  background: #141527;
  display: flex;
  height: 100vh;
  text-align: center;
  width: 100%;
}

#loader-container {
  height: 65vh;
}

.loader {
  animation: spin 2s linear infinite;
  border-bottom: 16px solid #1b1a1c;
  border-left: 16px solid #1b1a1c;
  border-radius: 50%;
  border-right: 16px solid #1b1a1c;
  border-top: 16px solid #489f01;
  box-shadow: 0px -5px 20px 3px #499f00;
  height: 120px;
  left: 45%;
  position: absolute;
  top: 45%;
  width: 120px;
}

@-webkit-keyframes spin {
  0% { -webkit-transform: rotate(0deg); }
  100% { -webkit-transform: rotate(360deg); }
}

@keyframes spin {
  0% { transform: rotate(0deg); }
  100% { transform: rotate(360deg); }
}
