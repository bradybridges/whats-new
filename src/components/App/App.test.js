import React from 'react';
import { shallow } from 'enzyme';
import App from './App';
import healthData from '../../data/health';

describe('App', () => {
  let wrapper;
  beforeEach(() => {
    wrapper = shallow(<App />);
    wrapper.instance().setState({ news: { health: healthData } });
  });

  it('should match snapshot', () => {
    expect(wrapper).toMatchSnapshot();
  });

  describe('updateCurrentCategory', () => {
    it('should be able to update currentCategory', () => {
      const expected = 'health';
      wrapper.instance().updateCurrentCategory('health');
      expect(wrapper.state('currentCategory')).toEqual(expected);
    });
  });

  describe('updateCurrentNews', () => {
    it('should be able to update current news', () => {
      const expected = 'health';
      wrapper.instance().updateCurrentCategory('health');
      expect(wrapper.state('currentNews')).toEqual(healthData);
    });
  });

  describe('updateNews', () => {
    it('should update currentNews based on a search value', () => {
      const closest = jest.fn().mockImplementation(() => {
        return { children: [{ value: 'man dies' }] };
      });

      const mockEvent = {
        target: {
          value: 'man dies',
          closest,
        },
        preventDefault: jest.fn(),
      };
      wrapper.instance().updateCurrentCategory('health');
      wrapper.instance().updateNews(mockEvent);
      expect(wrapper.state('currentNews')).toEqual([healthData[0]]);
    });
  });
});
