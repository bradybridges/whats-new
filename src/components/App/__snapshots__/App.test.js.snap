// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App should match snapshot 1`] = `
<main>
  <header>
    <h1>
      Denver News
    </h1>
    <SearchForm
      updateNews={[Function]}
    />
  </header>
  <Menu
    updateCurrentCategory={[Function]}
  />
  <section
    id="loader-container"
  >
    <div
      className="loader"
    />
  </section>
</main>
`;
